### ISSUE TITLE
Assertion failed: (!ctx->utp_sockets.contains(UTPSocketKey{ addr, conn_id_recv })), function ~UTPSocket, file utp_internal.cpp, line 2508.

### ISSUE THREAD

**Issue by Coeur** [transmission/transmission#6256]
### What is the issue?

When quitting Transmission.
>Assertion failed: (!ctx->utp_sockets.contains(UTPSocketKey{ addr, conn_id_recv })), function ~UTPSocket, file utp_internal.cpp, line 2508.
```
Thread 5#0	0x00007ff810af61e2 in __pthread_kill ()
#1	0x000000010c5765c2 in pthread_kill ()
#2	0x00007ff810a54b45 in abort ()
#3	0x00007ff810a53e5e in __assert_rtn ()
#4	0x000000010b19c36f in UTPSocket::~UTPSocket() at /Users/User/Development/transmission/third-party/libutp/utp_internal.cpp:2508
#5	0x000000010b19c8d9 in UTPSocket::~UTPSocket() at /Users/User/Development/transmission/third-party/libutp/utp_internal.cpp:2494
#6	0x000000010b19c937 in utp_socket_delete(UTPSocket*) at /Users/User/Development/transmission/third-party/libutp/utp_internal.cpp:2526
#7	0x000000010b18191b in UTPSocketHT::SocketDeleter::operator()(UTPSocket*) const at /Users/User/Development/transmission/third-party/libutp/utp_internal.h:123
#8	0x000000010b181729 in std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>::reset[abi:v160006](UTPSocket*) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/unique_ptr.h:297
#9	0x000000010b18156d in std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>::~unique_ptr[abi:v160006]() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/unique_ptr.h:263
#10	0x000000010b181519 in std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>::~unique_ptr[abi:v160006]() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/unique_ptr.h:263
#11	0x000000010b1814bd in std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>::~pair() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__utility/pair.h:63
#12	0x000000010b181469 in std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>::~pair() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__utility/pair.h:63
#13	0x000000010b18124d in void std::__1::allocator_traits<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>>>::destroy[abi:v160006]<std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void, void>(std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>>&, std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>*) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__memory/allocator_traits.h:325
#14	0x000000010b180efb in std::__1::__hash_table<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::__unordered_map_hasher<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::hash<UTPSocketKey>, std::__1::equal_to<UTPSocketKey>, true>, std::__1::__unordered_map_equal<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::equal_to<UTPSocketKey>, std::__1::hash<UTPSocketKey>, true>, std::__1::allocator<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>>>::__deallocate_node(std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>*) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__hash_table:1537
#15	0x000000010b180cf4 in std::__1::__hash_table<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::__unordered_map_hasher<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::hash<UTPSocketKey>, std::__1::equal_to<UTPSocketKey>, true>, std::__1::__unordered_map_equal<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::equal_to<UTPSocketKey>, std::__1::hash<UTPSocketKey>, true>, std::__1::allocator<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>>>::~__hash_table() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__hash_table:1478
#16	0x000000010b180bf9 in std::__1::__hash_table<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::__unordered_map_hasher<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::hash<UTPSocketKey>, std::__1::equal_to<UTPSocketKey>, true>, std::__1::__unordered_map_equal<UTPSocketKey, std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, std::__1::equal_to<UTPSocketKey>, std::__1::hash<UTPSocketKey>, true>, std::__1::allocator<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>>>::~__hash_table() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__hash_table:1470
#17	0x000000010b180ba9 in std::__1::unordered_map<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>, std::__1::hash<UTPSocketKey>, std::__1::equal_to<UTPSocketKey>, std::__1::allocator<std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>>>::~unordered_map[abi:v160006]() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/unordered_map:1153
#18	0x000000010b180b59 in std::__1::unordered_map<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>, std::__1::hash<UTPSocketKey>, std::__1::equal_to<UTPSocketKey>, std::__1::allocator<std::__1::pair<UTPSocketKey const, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>>>::~unordered_map[abi:v160006]() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/unordered_map:1151
#19	0x000000010b180b09 in UTPSocketHT::~UTPSocketHT() at /Users/User/Development/transmission/third-party/libutp/utp_internal.h:115
#20	0x000000010b180a19 in UTPSocketHT::~UTPSocketHT() at /Users/User/Development/transmission/third-party/libutp/utp_internal.h:115
#21	0x000000010b1809a0 in struct_utp_context::~struct_utp_context() at /Users/User/Development/transmission/third-party/libutp/utp_internal.h:161
#22	0x000000010b17f569 in struct_utp_context::~struct_utp_context() at /Users/User/Development/transmission/third-party/libutp/utp_internal.h:161
#23	0x000000010b17f505 in ::utp_destroy(utp_context *) at /Users/User/Development/transmission/third-party/libutp/utp_api.cpp:106
#24	0x000000010ae7746f in tr_utpClose(tr_session*) at /Users/User/Development/transmission/libtransmission/tr-utp.cc:226
#25	0x000000010a80b78c in tr_session::closeImplPart2(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>) at /Users/User/Development/transmission/libtransmission/session.cc:1400
#26	0x000000010a875d76 in tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6::operator()() const at /Users/User/Development/transmission/libtransmission/session.cc:1375
#27	0x000000010a875d19 in decltype(std::declval<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6&>()()) std::__1::__invoke[abi:v160006]<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6&>(tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6&) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/invoke.h:394
#28	0x000000010a875be9 in void std::__1::__invoke_void_return_wrapper<void, true>::__call<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6&>(tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6&) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/invoke.h:487
#29	0x000000010a875b95 in std::__1::__function::__alloc_func<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6, std::__1::allocator<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6>, void ()>::operator()[abi:v160006]() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/function.h:185
#30	0x000000010a872a11 in std::__1::__function::__func<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6, std::__1::allocator<tr_session::closeImplPart1(std::__1::promise<void>*, std::__1::chrono::time_point<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>)::$_6>, void ()>::operator()() at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/function.h:356
#31	0x000000010a9da10e in std::__1::__function::__value_func<void ()>::operator()[abi:v160006]() const at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/function.h:510
#32	0x000000010a9d9ff9 in std::__1::function<void ()>::operator()() const at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/function.h:1156
#33	0x000000010af5c31f in libtransmission::EvTimer::handleTimer() at /Users/User/Development/transmission/libtransmission/timer-ev.cc:151
#34	0x000000010af5b973 in libtransmission::EvTimer::onTimer(int, short, void*) at /Users/User/Development/transmission/libtransmission/timer-ev.cc:143
#35	0x000000010b0b3aac in event_persist_closure at /Users/User/Development/transmission/third-party/libevent/event.c:1623
#36	0x000000010b0aff63 in event_process_active_single_queue at /Users/User/Development/transmission/third-party/libevent/event.c:1682
#37	0x000000010b095b8b in event_process_active at /Users/User/Development/transmission/third-party/libevent/event.c:1783
#38	0x000000010b0903cd in event_base_loop at /Users/User/Development/transmission/third-party/libevent/event.c:2006
#39	0x000000010aa22652 in tr_session_thread_impl::sessionThreadFunc(event_base*) at /Users/User/Development/transmission/libtransmission/session-thread.cc:239
#40	0x000000010aa26a38 in decltype(*std::declval<tr_session_thread_impl*>().*std::declval<void (tr_session_thread_impl::*)(event_base*)>()(std::declval<event_base*>())) std::__1::__invoke[abi:v160006]<void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*, void>(void (tr_session_thread_impl::*&&)(event_base*), tr_session_thread_impl*&&, event_base*&&) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__functional/invoke.h:359
#41	0x000000010aa2697d in void std::__1::__thread_execute[abi:v160006]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*, 2ul, 3ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*>&, std::__1::__tuple_indices<2ul, 3ul>) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/thread:288
#42	0x000000010aa25c7e in void* std::__1::__thread_proxy[abi:v160006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*>>(void*) at /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/thread:299
#43	0x000000010c5757eb in _pthread_start ()
#44	0x000000010c57dbef in thread_start ()
```
<img width="1792" alt="Screenshot 2023-11-14 at 23 12 16" src="https://github.com/transmission/transmission/assets/839992/3e15391c-b97a-4f29-98b9-cb8bc30d2077">

### Which application of Transmission?

macOS app

### Which version of Transmission?

cf077bbba08a7da7c85629a2a635211650ec15a9

**Comment by ckerr** [comment]
_sigh_

**Comment by ckerr** [comment]
Fixed in 7c177224ed06cb5ea2078e16ffe70a3fb2341158.

Looks like the lifecycle management code is OK, but it's not a great idea to call `hashtable.contains()` while we're in the middle of the hashtable's destructor :smile:
